pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,993 @@
+	} else {
+		curState.listenPort = defaultBGPListenPort
+	}
+	curState.localASN = effectiveBP.Spec.LocalASN
+
+	// If any of the listen port, local AS number, or router ID have changed, a new BGP server should be also started.
+	if preState != nil {
+		needUpdateBGPServer = needUpdateBGPServer ||
+			preState.listenPort != curState.listenPort ||
+			preState.localASN != curState.localASN ||
+			preState.routerID != curState.routerID
+	}
+
+	var success bool
+	if needUpdateBGPServer {
+		// Stop the stale BGP server.
+		if preState != nil {
+			if err := preState.bgpServer.Stop(c.ctx); err != nil {
+				return fmt.Errorf("failed to stop stale BGP server: %w", err)
+			}
+		}
+
+		// Get the new BGP server.
+		globalConfig := &bgp.GlobalConfig{
+			ASN:        uint32(curState.localASN),
+			RouterID:   curState.routerID,
+			ListenPort: curState.listenPort,
+		}
+		bgpServer := c.newBGPServerFn(globalConfig)
+
+		// Start the new BGP server.
+		if err := bgpServer.Start(c.ctx); err != nil {
+			return fmt.Errorf("failed to start BGP server: %w", err)
+		}
+
+		// Stop the new BGP server if any failure occurs in later manipulation.
+		defer func() {
+			if !success {
+				bgpServer.Stop(c.ctx)
+			}
+		}()
+
+		curState.bgpServer = bgpServer
+	} else {
+		curState.bgpServer = preState.bgpServer
+	}
+
+	// Reconcile BGP peers.
+	curPeerConfigs, err := c.getPeerConfigs(effectiveBP.Spec.BGPPeers)
+	if err != nil {
+		return err
+	}
+	prePeerConfigs := make(map[string]bgp.PeerConfig)
+	if preState != nil {
+		prePeerConfigs = preState.peerConfigs
+	}
+	if err := c.reconcileBGPPeers(curPeerConfigs, prePeerConfigs, curState.bgpServer, needUpdateBGPServer); err != nil {
+		return err
+	}
+
+	// Reconcile BGP routes generated from advertisements.
+	curRoutes, err := c.getRoutes(effectiveBP.Spec.Advertisements)
+	if err != nil {
+		return err
+	}
+	preRoutes := sets.Set[bgp.Route]{}
+	if preState != nil {
+		preRoutes = preState.routes
+	}
+	if err := c.reconcileRoutes(curRoutes, preRoutes, curState.bgpServer, needUpdateBGPServer); err != nil {
+		return err
+	}
+
+	// Update the current state.
+	curState.routes = curRoutes
+	curState.peerConfigs = curPeerConfigs
+	c.bgpPolicyState = curState
+
+	success = true
+
+	return nil
+}
+
+func getPeerConfigsForKeys(peerKeys sets.Set[string], allPeerConfigs map[string]bgp.PeerConfig) []bgp.PeerConfig {
+	peerConfigs := make([]bgp.PeerConfig, 0, len(peerKeys))
+	for peer := range peerKeys {
+		peerConfigs = append(peerConfigs, allPeerConfigs[peer])
+	}
+	return peerConfigs
+}
+
+func (c *Controller) reconcileBGPPeers(curPeerConfigs, prePeerConfigs map[string]bgp.PeerConfig, bgpServer bgp.Interface, bgpServerUpdated bool) error {
+	prePeerKeys := sets.KeySet(prePeerConfigs)
+	curPeerKeys := sets.KeySet(curPeerConfigs)
+
+	var peerToAddKeys sets.Set[string]
+	if !bgpServerUpdated {
+		peerToAddKeys = curPeerKeys.Difference(prePeerKeys)
+	} else {
+		peerToAddKeys = curPeerKeys
+	}
```

Comment:
If the cache is up-to-date, this handling is not needed.

I'm not sure if the current code is correct if stopping/starting bgp server fails in one round and succeeds next round. But I think it's more clear if we always keep cache up-to-date.

Reply:
I have updated the code here, but I'm not sure if it is what you expected.